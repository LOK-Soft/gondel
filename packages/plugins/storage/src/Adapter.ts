export interface IStorage {
  readonly length: number;
  clear(): void;
  getItem(key: string): string | null;
  key(index: number): string | null;
  removeItem(key: string): void;
  setItem(key: string, value: string): void;
  [name: string]: any;
}

export interface ISerializer<T extends any = any> {
  serialize: (value: T) => string;
  deserialize: (value: string) => T;
}

type AdditionalAdapterConfiguration = {
  prefix: string;
  postfix: string;
  delimitier: string;
};

export class Adapter {
  private config: AdditionalAdapterConfiguration = {
    prefix: "",
    postfix: "",
    delimitier: ""
  };

  constructor(private name: string, private store: IStorage) {
    if (!store) {
      throw new Error(`Stroage not available in your environment`);
    }
  }

  configure(config: AdditionalAdapterConfiguration) {
    this.config = {
      ...this.config,
      ...config
    };
  }

  get<T, S extends ISerializer<T>>(key: string, serializer?: S): T | void {
    const value = this.store.getItem(this.generateAccessorKey(key));

    if (value) {
      if (serializer) {
        return serializer.deserialize(value);
      }

      return (value as unknown) as T;
    }

    return undefined;
  }

  set<T extends any = string>(key: string, value: T, serializer?: ISerializer<T>): void {
    let savableValue: any = "";

    if (serializer && typeof value !== "string") {
      savableValue = serializer.serialize(value);
    } else {
      savableValue = value;
    }

    this.store.setItem(this.generateAccessorKey(key), savableValue);
  }

  remove(key: string): void {
    this.store.removeItem(key);
  }

  clear(): void {
    this.store.clear();
  }

  toString(): string {
    return `Adapter { ${this.name} }`;
  }

  private generateAccessorKey(key: string): string {
    return [this.config.prefix, key, this.config.postfix].join(this.config.delimitier);
  }
}

export const localStorageAdapter = new Adapter("local", window.localStorage as IStorage);
export const sessionStorageAdapter = new Adapter("session", window.sessionStorage as IStorage);
