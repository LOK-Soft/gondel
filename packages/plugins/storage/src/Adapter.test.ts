import { localStorageAdapter, sessionStorageAdapter, Adapter } from "./Adapter";
import JSONSerializer from "./serializer/JSON";

beforeEach(() => {
  localStorageAdapter.clear();
  sessionStorageAdapter.clear();
});

describe("@gondel/plugin-storage", () => {
  describe("Adapter", () => {
    it("should expose members", () => {
      expect(localStorageAdapter).toBeDefined();
      expect(sessionStorageAdapter).toBeDefined();
    });

    it("should fail if no env given", () => {
      expect(() => {
        // test case for no localStorage (e.g. old browsers)
        new Adapter("noEnvGiven", undefined as any);
      }).toThrowError();
    });

    it("should stringify correctly", () => {
      expect(localStorageAdapter.toString()).toEqual("Adapter { local }");
      expect(sessionStorageAdapter.toString()).toEqual("Adapter { session }");
    });

    it("should get/set simple items correctly", () => {
      localStorageAdapter.set("test1", "test1");
      expect(JSON.stringify(localStorage)).toMatchSnapshot();
      expect(localStorageAdapter.get("test1")).toEqual("test1");
    });

    it("should get/set custom serializable items correctly", () => {
      localStorageAdapter.set("test2", { a: 10, b: 20 }, JSONSerializer);
      expect(JSON.stringify(localStorage)).toMatchSnapshot();
      expect(localStorageAdapter.get("test2", JSONSerializer)).toEqual({
        a: 10,
        b: 20
      });
    });

    it("should remove items", () => {
      localStorageAdapter.set("test3", "remove");
      localStorageAdapter.remove("test3");
      expect(localStorageAdapter.get("test3")).toBeUndefined();
    });
  });
});
