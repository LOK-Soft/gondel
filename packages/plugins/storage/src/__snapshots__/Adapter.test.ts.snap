// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`@gondel/plugin-storage Adapter should get/set custom serializable items correctly 1`] = `"{\\"test2\\":\\"{\\\\\\"a\\\\\\":10,\\\\\\"b\\\\\\":20}\\"}"`;

exports[`@gondel/plugin-storage Adapter should get/set simple items correctly 1`] = `"{\\"test1\\":\\"test1\\"}"`;
